approxFutureEnergy.m

```diff
@@ -1,131 +1,176 @@
-  end
+if (nargin < 7)
+  verbose = false;
+end
 
-  n = size(A,1);   % A should be n-by-n
-  m = size(B,2);   % B should be n-by-m
-  p = size(C,1);   % C should be p-by-n
+% Create pointer/shortcuts for dynamical system polynomial coefficients
+if ismatrix(g)
+  % Will reduce to Jeff's original code
+  B = g;
+  l = 0;
+  g = {B};
+else
+  % QB or polynomial input balancing
+  B = g{1};
+  l = length(g) - 1;
+end
 
-  % Create a vec function for readability
-  vec = @(X) X(:);
+n = size(A, 1); % A should be n-by-n
+m = size(B, 2); % B should be n-by-m
+p = size(C, 1); % C should be p-by-n
 
+% Create a vec function for readability
+vec = @(X) X(:);
 
-  %% k=2 case
-  R = eye(m)/eta;
+%% k=2 case
+R = eye(m) / eta;
 
-  if ( eta>0 )
-    [W2] = icare(A,B,(C.'*C),R);
-    if ( isempty(W2) && verbose )
-      warning('approxFutureEnergy: icare couldn''t find stabilizing solution')
-    end
-    
-  elseif ( eta<0 )
-    [W2] = icare(A,B,(C.'*C),R,'anti');
-    
-    if ( isempty(W2) && verbose )
-      warning('approxFutureEnergy: icare couldn''t find stabilizing solution')
-      warning('approxFutureEnergy: using the hamiltonian')
-      [~,W2,~] = hamiltonian(A,B,C.'*C,R,true);
-    end
-    
-  else % eta==0
-    [W2] = lyap(A.',(C.'*C));
-    
-  end
+if (eta > 0)
+  [W2] = icare(A, B, (C.' * C), R);
 
-  if (isempty(W2))
-    error('approxFutureEnergy: Can''t find a stabilizing solution')
+  if (isempty(W2) && verbose)
+    warning('approxFutureEnergy: icare couldn''t find stabilizing solution')
   end
-  
-  %  Check the residual of the Riccati/Lyapunov equation
-  if (verbose)
-    RES = A'*W2 + W2*A - eta*(W2*B)*(B'*W2) + C'*C ;
-    fprintf('The residual of the Riccati equation is %g\n',norm(RES,'inf'));
+
+elseif (eta < 0)
+  [W2] = icare(A, B, (C.' * C), R, 'anti');
+
+  if (isempty(W2) && verbose)
+    warning('approxFutureEnergy: icare couldn''t find stabilizing solution')
+    warning('approxFutureEnergy: using the hamiltonian')
+    [~, W2, ~] = hamiltonian(A, B, C.' * C, R, true);
   end
 
-  %  Reshape the resulting quadratic coefficients
-  w2 = W2(:);
-  w{2} = w2;
+else % eta==0
+  [W2] = lyap(A.', (C.' * C));
 
-  %% k=3 case
-  if ( d>2 ) % set up the generalized Lyapunov solver
-    W2BB = W2*(B*B.');
-    Acell = cell(1,d);
-    for i=1:d
-      Acell{i} = A.'-eta*W2BB;
-    end
+end
+
+if (isempty(W2))
+  error('approxFutureEnergy: Can''t find a stabilizing solution')
+end
+
+%  Check the residual of the Riccati/Lyapunov equation
+if (verbose)
+  RES = A' * W2 + W2 * A - eta * (W2 * B) * (B' * W2) + C' * C;
+  fprintf('The residual of the Riccati equation is %g\n', norm(RES, 'inf'));
+end
+
+%  Reshape the resulting quadratic coefficients
+w{2} = vec(W2);
+
+%% k=3 case
+if (d > 2)
+  GaWb = cell(2 * l + 1, d - 1); % Pre-compute G_a.'*W_b, etc for all the a,b we need
+  GaWb{1, 2} = B.' * W2;
+  Im = speye(m);
+  % set up the generalized Lyapunov solver
+  [Acell{1:d}] = deal(A.' - eta * W2 * (B * B.'));
```

Comment:
I use `deal` rather than a for loop because it allocates memory once rather than $d$ times, since all the entries in `Acell` are the same.